packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -123,9 +123,81 @@ class LineBoundary extends TextBoundary {
   TextRange getTextBoundaryAt(int position) => _textLayout.getLineAtOffset(TextPosition(offset: max(position, 0)));
 }
 
+/// A text boundary that uses paragraphs as logical boundaries.
+///
+/// A paragraph is defined as the range between line terminators. If no
+/// line terminators exist then the paragraph boundary is the entire document.
+class ParagraphBoundary extends TextBoundary {
+  /// Creates a [ParagraphBoundary] with the text.
+  const ParagraphBoundary(this._text);
+
+  final String _text;
+
+  /// Returns the [int] representing the start position of the paragraph that
+  /// bounds the given `position`. The returned [int] is at the front of the leading
+  /// line terminator that encloses the desired paragraph.
+  @override
+  int? getLeadingTextBoundaryAt(int position) {
+    final List<int> codeUnits = _text.codeUnits;
```

Comment:
Would this work:

```dart
if (position < 0) { 
  return null;
} else if (position >= _text.length) { 
  return getLeadingTextBoundaryAt(_text.length - 1);
}

int index = position;
while (index > 0 && !isNewline(codeUnits[index])) { 
  index -= 1;
}
```

You should be able to implement `getTrailingTextBoundaryAt` in a similar fashion, with an additional look ahead step when `index` points to a `CR`.